storage/src/main/java/tech/pegasys/teku/storage/store/Store.java

```diff
@@ -597,4 +676,21 @@ private void putBlock(final SignedBeaconBlock block) {
       writeLock.unlock();
     }
   }
+
+  Bytes32 applyForkChoiceScoreChanges(
+      final MutableStore transaction,
+      final Checkpoint finalizedCheckpoint,
+      final Checkpoint justifiedCheckpoint,
+      final BeaconState justifiedCheckpointState) {
+    // Ensure the store lock is taken before entering forkChoiceStrategy. Otherwise it takes the
+    // protoArray lock first, and may deadlock when it later needs to get votes which requires the
+    // store lock.
```

Comment:
To make sure I understand, the deadlock would happen in the following way?:
* We enter this method, run `findHead` and acquire the protoarray locks
* Another transaction finishes, acquires store locks, and tries to acquire protoarray locks in `BlockMetadataStore.applyUpdates`, but can't
* `findHead` now tries to acquire the store locks but can't

Reply:
And `ProtoArrayForkChoiceStrategy.onAttestation` doesn't create a similar problem because it's only acquiring the `votesLock`?

Reply:
Correct on both counts. The deadlock is between two transactions, one that's running through `findHead` and one that's importing blocks.